### PR TITLE
feat: restricts unverified user for buying free tickets

### LINKED ISSUE #6034: Don't allow unverified users to place free orders
Unverified users shouldn't be allowed to place free orders, paid orders are okay.

### PR DESCRIPTION
**Pull request by prateekj117**
Fixes #6034

#### Short description of what this resolves:
It adds a check to restrict an unverified user from placing free order.

#### Checklist

- [X] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [X] My branch is up-to-date with the Upstream `development` branch.
- [X] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)

- [X] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

**Comment by prateekj117** [comment]
@iamareebjamal Please review.

File: app/api/attendees.py [hunk 1]
```diff
@@ -170,18 +170,18 @@ def before_update_object(self, obj, data, kwargs):
...
                 raise UnprocessableEntity(
-                    {'pointer': '/data/relationships/ticket'}, "Invalid Ticket"
+                    {'pointer': '/data/relationships/order'}, "Invalid Order"
                 )
-            if not user.is_verified and ticket.price == 0:
+            if not user.is_verified and order.amount == 0:
                 raise UnprocessableEntity(
-                    {'pointer': '/data/relationships/ticket'},
+                    {'pointer': '/data/relationships/order'},
                     "Unverified user cannot buy free tickets"
```
Comment (kushthedude) [inline app/api/attendees.py]: `Unverified user cant place free orders` @prateekj117

File: app/api/attendees.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 from datetime import datetime
 
 from flask import Blueprint, request, jsonify, abort, make_response
-from flask_jwt import current_identity, jwt_required
+from flask_jwt import current_identity
```
Comment (kushthedude) [inline app/api/attendees.py]: Why did you change this?

**Comment by prateekj117** [comment]
@kushthedude This check is before the discount code is applied.

**Comment by kushthedude** [comment]
> @kushthedude This check is before the discount code is applied.

Nope, I can't place free order with the discount code, Please check again .

**Comment by kushthedude** [comment]
> #### Short description of what this resolves:

@prateekj117 Please fill the template and if not at least fill the issue number it fixes, Makes it easy to automatically close the issue

File: app/api/attendees.py (hunk 2 above)
Comment (prateekj117) [reply app/api/attendees.py]: @kushthedude Was not used anywhere in file, so I guess Pycharm did it.

File: app/api/attendees.py [hunk 3]
```diff
@@ -170,18 +170,18 @@ def before_update_object(self, obj, data, kwargs):
...
+                id=int(data['order']), deleted_at=None
                 ).first()
-            if ticket is None:
+            if order is None:
                 raise UnprocessableEntity(
-                    {'pointer': '/data/relationships/ticket'}, "Invalid Ticket"
+                    {'pointer': '/data/relationships/order'}, "Invalid Order"
                 )
-            if not user.is_verified and ticket.price == 0:
+            if not user.is_verified and order.amount == 0:
```
Comment (mrsaicharan1) [inline app/api/attendees.py]: You have to check with the `type` of the order.

**Comment by prateekj117** [comment]
@iamareebjamal Please review.

File: app/api/attendees.py [hunk 4]
```diff
@@ -170,21 +170,6 @@ def before_update_object(self, obj, data, kwargs):
...
-                ).first()
-            if ticket is None:
-                raise UnprocessableEntity(
-                    {'pointer': '/data/relationships/ticket'}, "Invalid Ticket"
-                )
-            if not user.is_verified and ticket.price == 0:
-                raise UnprocessableEntity(
-                    {'pointer': '/data/relationships/ticket'},
-                    "Unverified user cannot buy free tickets"
-                )
```
Comment (iamareebjamal) [inline app/api/attendees.py]: Why is this removed?

File: app/api/orders.py [hunk 5]
```diff
@@ -56,6 +55,12 @@ def before_post(self, args, kwargs, data=None):
         """
         require_relationship(['event'], data)
 
+        if not current_user.is_verified and data['payment_mode'] == 'free':
+            raise UnprocessableEntity(
+                {'pointer': '/data/relationships/order'},
+                "Unverified user cannot place free orders"
+            )
```
Comment (iamareebjamal) [inline app/api/orders.py]: I can still select free ticket and send payment_mode as something else. Boom

**Comment by prateekj117** [comment]
@iamareebjamal Please review.

File: app/api/orders.py [hunk 6]
```diff
@@ -86,6 +87,16 @@ def before_create_object(self, data, view_kwargs):
                 raise ConflictException({'pointer': '/data/relationships/attendees'},
                                         "Attendee with id {} does not exists".format(str(ticket_holder)))
 
+        ticket = db.session.query(Ticket).filter_by(
+            id=int(ticket_holder_object.ticket_id), deleted_at=None
+        ).first()
```
Comment (mrsaicharan1) [inline app/api/orders.py]: What about an order with free & paid tickets for unverified users? They should be blocked as well too right?

**Comment by kushthedude** [comment]
No they shouldnt , and also change the title of PR yo just free orders ,
not free tickets!

File: app/api/orders.py (hunk 6 above)
Comment (iamareebjamal) [inline app/api/orders.py]: This only checks for last ticket holder. Please test your PRs before asking for reviews

File: app/api/orders.py [hunk 7]
```diff
@@ -86,6 +87,16 @@ def before_create_object(self, data, view_kwargs):
                 raise ConflictException({'pointer': '/data/relationships/attendees'},
                                         "Attendee with id {} does not exists".format(str(ticket_holder)))
 
+        ticket = db.session.query(Ticket).filter_by(
```
Comment (shreyanshdwivedi) [inline app/api/orders.py]: Okay, I think you are confused about what we are trying to achieve here. We have already implemented restriction on orders which have at least one free ticket placed by an unverified user. But we want to restrict the **orders** which are completely free (not only one ticket free). You are still restricting orders which have even a single free ticket and other paid tickets

**Comment by prateekj117** [comment]
@shreyanshdwivedi @uds5501 Please review.

File: app/api/orders.py [hunk 8]
```diff
@@ -74,6 +74,9 @@ def before_create_object(self, data, view_kwargs):
         :param view_kwargs:
         :return:
         """
+
+        free_ticket_quantity = 0
```
Comment (iamareebjamal) [inline app/api/orders.py]: Why is this variable declared so far from its use?

File: app/api/orders.py [hunk 9]
```diff
@@ -86,6 +89,19 @@ def before_create_object(self, data, view_kwargs):
                 raise ConflictException({'pointer': '/data/relationships/attendees'},
                                         "Attendee with id {} does not exists".format(str(ticket_holder)))
 
+            ticket = db.session.query(Ticket).filter_by(
+                id=int(ticket_holder_object.ticket_id), deleted_at=None
+            ).first()
```
Comment (iamareebjamal) [inline app/api/orders.py]: Why doing all this and not just `ticket_holder_object.ticket`?

File: app/api/orders.py [hunk 10]
```diff
@@ -86,6 +89,19 @@ def before_create_object(self, data, view_kwargs):
...
+
+            if not current_user.is_verified and ticket.type == 'free':
```
Comment (iamareebjamal) [inline app/api/orders.py]: No need to check `not current_user.is_verified` in a loop, it is loop invariant

File: app/api/orders.py [hunk 11]
```diff
@@ -86,6 +89,19 @@ def before_create_object(self, data, view_kwargs):
...
+                free_ticket_quantity += 1
+
+        if free_ticket_quantity == len(data['ticket_holders']):
```
Comment (iamareebjamal) [inline app/api/orders.py]: Check `not current_user.is_verified` here only

**Comment by iamareebjamal** [comment]
Looks fine to me, ask peers to review. Taking 8 days for a PR this trivial is not expected

**Comment by prateekj117** [comment]
@shreyanshdwivedi @uds5501 @mrsaicharan1 Please review.
